pkg/controller/virtualnode.go

```diff
@@ -296,3 +337,79 @@ func (c *Controller) handleVNodeMeshDeleting(ctx context.Context, vnode *appmesh
 	}
 	return true
 }
+
+func (c *Controller) handleServiceDiscovery(ctx context.Context, vnode *appmeshv1beta1.VirtualNode, copyForUpdate *appmeshv1beta1.VirtualNode) error {
+	//only handle cloudmap for now
+	if vnode.Spec.ServiceDiscovery == nil || vnode.Spec.ServiceDiscovery.CloudMap == nil {
+		klog.V(4).Infof("CloudMap service is already created")
+		return nil
+	}
+
+	if vnode.Status.CloudMapService != nil && vnode.Status.CloudMapService.ServiceID != nil {
+		klog.V(4).Infof("CloudMap service is already created for virtual-node %s", vnode.Name)
+		return nil
+	}
+
+	cloudmapNamespaceName := vnode.Spec.ServiceDiscovery.CloudMap.NamespaceName
+	cloudmapServiceName := vnode.Spec.ServiceDiscovery.CloudMap.ServiceName
+
+	if cloudmapNamespaceName == "" {
+		return fmt.Errorf("CloudMap servicediscovery is missing NamespaceName value for virtual node %s", vnode.Name)
+	}
+
+	if cloudmapServiceName == "" {
+		return fmt.Errorf("CloudMap servicediscovery is missing ServiceName value for virtual node %s", vnode.Name)
+	}
+
+	cloudmapConfig := &appmesh.AwsCloudMapServiceDiscovery{
+		NamespaceName: awssdk.String(cloudmapNamespaceName),
+		ServiceName:   awssdk.String(cloudmapServiceName),
+	}
+
+	//It is okay to call Create multiple times for same service-name.
+	//It is also cheaper than calling get and then figuring out to create.
+	cloudmapService, err := c.cloud.CloudMapCreateService(ctx, cloudmapConfig, c.name)
```

Comment:
Do we need to handle the case where the cloudmap namespace doesn't exist?  I guess its probably evident with the error returned from the cloudmap create service call, which is already logged.

Reply:
we probably should but creating namespace on the fly seems bad. CloudMap namespace should be created by admin, same can be said of mesh though. For now I will add a TODO to improve the experience.